horovod/tensorflow/__init__.py

```diff
@@ -252,15 +371,18 @@ def compute_gradients(self, *args, **kwargs):
             """
             gradients = self._optimizer.compute_gradients(*args, **kwargs)
             if size() > 1:
-                grads, vars = zip(*gradients)
-                avg_grads = self._allreduce_grads(grads)
-                return list(zip(avg_grads, vars))
+                self.grads, vars = zip(*gradients)
+                self._agg_helper.init_aggregation_vars(self.grads)
+                allreduced_grads = self._agg_helper.compute_gradients(self.grads)
+                return list(zip(allreduced_grads, vars))
             else:
                 return gradients
 
         def apply_gradients(self, *args, **kwargs):
             """Calls this same method on the underlying optimizer."""
-            return self._optimizer.apply_gradients(*args, **kwargs)
+            flattended_args0 = [item for tup in args[0] for item in tup]
+            with tf.control_dependencies([tf.group(*flattended_args0)]):
```

Comment:
Yeah I was thinking about that.. we have two ways I can think of

1) Give `LocalGradientAggregationHelper` access to the optimizer. But this might go counter to the goals I had of it being something that works for both TF Keras and Tensorflow. More exploration may be needed. But it would look something like:
```
def apply_gradients(self, optimizer, *args, **kwargs):
    flattended_args0 = [item for tup in args[0] for item in tup]
    with tf.control_dependencies([tf.group(*flattended_args0)]):
        return tf.cond(tf.equal(self.counter, 0), lambda: optimizer.apply_gradients(*args, **kwargs), tf.no_op)
```
2) Have a function that looks like 
```
def apply_gradients(self, apply_grads_func_closure, *args, **kwargs):
    flattended_args0 = [item for tup in args[0] for item in tup]
    with tf.control_dependencies([tf.group(*flattended_args0)]):
        return tf.cond(tf.equal(self.counter, 0), apply_grads_func_closure, tf.no_op)
```

thoughts?

Reply:
Yeah I think second approach is better, as it's best to avoid passing `self._optimizer` into a function call.